File: 0078-subsets/0078-subsets.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums) {
    let res = [[]];
    const backtrack = (start, cur) =>{
        for(let i = start; i < nums.length; i++){
            cur.push(nums[i]);
            res.push([...cur])
            backtrack(i+1, cur);
            cur.pop();
            // console.log(cur, i);
            // console.log(res, i);
        }
    }
    backtrack(0,[]);
    return res;
};

File: 0078-subsets/0078-subsets.java
class Solution {
    public List<List<Integer>> res = new ArrayList<>();
    
    public List<List<Integer>> subsets(int[] nums) {
        backtrack(0, new ArrayList<>(), nums);
        return res;
    }
    
    private void backtrack(int start, List<Integer> cur, int[] nums){
        res.add(new ArrayList<>(cur));
        for(int i = start; i < nums.length; i++){
            cur.add(nums[i]);
            backtrack(i+1, cur, nums);
            cur.remove(cur.size() - 1);
        }
    }
}